src/nm.c

```diff
@@ -254,6 +254,61 @@ write_routes_nm(const NetplanNetDefinition* def, GKeyFile *kf, gint family, GErr
             j++;
         }
     }
+
+    return TRUE;
+}
+
+STATIC gboolean
+write_ip_rules_nm(const NetplanNetDefinition* def, GKeyFile *kf, gint family, GError** error)
+{
+    const gchar* group = NULL;
+    gchar* tmp_key = NULL;
+    GString* tmp_val = NULL;
+
+    if (family == AF_INET)
+        group = "ipv4";
+    else if (family == AF_INET6)
+        group = "ipv6";
+    g_assert(group != NULL);
+
+    if (def->ip_rules != NULL) {
+        for (unsigned i = 0, j = 1; i < def->ip_rules->len; ++i) {
+            const NetplanIPRule *cur_rule = g_array_index(def->ip_rules, NetplanIPRule*, i);
+
+            if (cur_rule->family != family)
+                continue;
+
+            /* NetworkManager requires that priority be specified */
+            if (cur_rule->priority == NETPLAN_IP_RULE_PRIO_UNSPEC) {
+                g_set_error(error, NETPLAN_BACKEND_ERROR, NETPLAN_ERROR_UNSUPPORTED,
+                            "ERROR: %s: priority not specified for routing policy\n", def->id);
```

Comment:
The ip-rule docs seem to have a strong opinion that a unique priority should always be specified for each rule.  For example, http://www.policyrouting.org/iproute2.doc.html#ss9.6.1 mentions `Each rule should have an explicitly set unique priority value` and has the following warning:

`For historical reasons ip rule add does not require any priority value and allows the priority value to be non-unique. If the user had not supplied a priority value then one was assigned by the kernel. If the user requested creating a rule with a priority value which already existed then the kernel did not reject the request and added the new rule before all old rules of the same priority. This is a mistake in the current design, nothing more. It should be fixed by the time you read this so please do not rely on this feature. You should always use explicit priorities when creating rules.`

Do you know if the sentiment among iproute2 devs has changed since that was written?

> Or should we rather set a default value here and log a debug message about it? But what would be a good default?

Not sure what a good default would be. Plus, to add an additional complication, how should a netplan automatically assigned priority behave in the presence of pass-through NetworkManager routing-ruleXs? I've only just become familiar with iproute2 (finally being forced to abandon ifconfig/route etc) so I'm not sure whether the kernel/systemd behavior of automatically assigned a priority has value in practice.

> Erroring out is probably fine... But could you add a XXX comment about maybe setting a default value there (and logging a debug message about it), in the future?

Let me know what you think about what I've written above and if you still think a comment is appropriate.

> Could you please update the "routing-policy.priority" docs in doc/netplan.yaml, adding a sentence like "This setting is mandatory on the NetworkManager back end."

Will do.

Reply:
> > Erroring out is probably fine... But could you add a XXX comment about maybe setting a default value there (and logging a debug message about it), in the future?
> 
> Let me know what you think about what I've written above and if you still think a comment is appropriate.

Thanks for digging up that documentation! I think a comment is still appropriate here, but instead of XXXing a future dynamic assignment, it should reference the documentation from above, to explain that this is the way we want it to be. We could even consider adding a similar check to `src/validation.c`, printing a warning when a rule is defined without explicit priority (no hard failure in `validation.c`, though, as that would break backwards compatibility).